src/imitation/algorithms/adversarial.py

```diff
@@ -147,65 +160,104 @@ def gen_policy(self) -> BaseRLModel:
+    if total_timesteps is None:
+      total_timesteps = self.gen_batch_size
+
     with logger.accumulate_means("gen"):
-      self._gen_policy.set_env(self.venv_train_norm)
+      self.gen_policy.set_env(self.venv_train_norm)
       # TODO(adam): learn was not intended to be called for each training batch
       # It should work, but might incur unnecessary overhead: e.g. in PPO2
       # a new Runner instance is created each time. Also a hotspot for errors:
       # algorithms not tested for this use case, may reset state accidentally.
-      self._gen_policy.learn(n_steps, reset_num_timesteps=False)
+      self.gen_policy.learn(total_timesteps=total_timesteps,
+                            reset_num_timesteps=False,
+                            callback=callback)
       self._populate_gen_replay_buffer()
 
+  def train(self, total_timesteps: int) -> None:
+    """Trains the discriminator and generator against each other.
+
+    Equivalent to calling `self.train_disc()` followed by `self.train_gen()`
+    in a loop `total_timesteps // self.gen_batch_size` times.
+
+    Args:
+        total_timesteps: The number of transitions to sample from
+          `self.venv_train_norm` during generator training.
+    """
+    n_epochs = total_timesteps // self.gen_batch_size
+    for _ in tqdm(range(n_epochs), desc="Adversarial train"):
+      self.train_disc()
+      self.train_gen()
+
   def _populate_gen_replay_buffer(self) -> None:
     """Generate and store generator samples in the buffer.
 
     More specifically, rolls out generator-policy trajectories in the
-    environment until `self._n_disc_samples_per_buffer` obs-act-obs samples are
+    environment until `self.disc_batch_size` obs-act-obs samples are
     produced, and then stores these samples.
     """
-    gen_rollouts = util.rollout.generate_transitions(
-        self._gen_policy, self.venv_train,
-        n_timesteps=self._n_disc_samples_per_buffer)
-    self._gen_replay_buffer.store(gen_rollouts)
-
-  def train(self, n_epochs=100, *, n_gen_steps_per_epoch=None,
-            n_disc_steps_per_epoch=None):
-    """Trains the discriminator and generator against each other.
-
-    Args:
-        n_epochs (int): The number of epochs to train. Every epoch consists
-            of training the discriminator and then training the generator.
-        n_disc_steps_per_epoch (int): The number of steps to train the
-            discriminator every epoch. More precisely, the number of full batch
-            Adam optimizer steps to perform.
-        n_gen_steps_per_epoch (int): The number of generator training steps
-            during each epoch. (ie, the timesteps argument in in
-            `policy.learn(timesteps)`).
-    """
-    for i in tqdm(range(n_epochs), desc="AIRL train"):
-      self.train_disc(**_n_steps_if_not_none(n_disc_steps_per_epoch))
-      self.train_gen(**_n_steps_if_not_none(n_gen_steps_per_epoch))
-      logger.dumpkvs()
+    gen_samples = util.rollout.generate_transitions(
```

Comment:
Is this OK in terms of not resetting the environment?

Reply:
No, but this environment reset is resolved by #144